spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/DataConverter.scala

```diff
@@ -0,0 +1,654 @@
+ * @param attributeBuilder  the attribute value writable instance
+ * @param value the value to be written to protobuf file
+ * @param valueType the type of the value to help set the data type
+ */
+  def setAttributeValue[T : ClassTag](attributeBuilder : AttrValue.Builder, value: Any,
+                                      valueType : universe.Type = null)
+    (implicit ev: TensorNumeric[T]) : Unit
+}
+
+/**
+ * General implementation of [[DataConverter]], it provides the conversion entry for all types
+ */
+object DataConverter extends DataConverter{
+
+  private val typePlaceHolder = universe.typeOf[DataConverter]
+
+  private def getRuntimeType[T : ClassTag](value : Any) (implicit ev: TensorNumeric[T])
+    : universe.Type = {
+    if (value.isInstanceOf[Tensor[T]]) {
+      ModuleSerializer.tensorType
+    } else if (value.isInstanceOf[AbstractModule[Activity, Activity, T]]) {
+      ModuleSerializer.abstractModuleType
+    } else if (value.isInstanceOf[Regularizer[T]]) {
+      ModuleSerializer.regularizerType
+    } else if (value.isInstanceOf[InitializationMethod]) {
+      universe.typeOf[InitializationMethod]
+    } else if (value.isInstanceOf[VariableFormat]) {
+      universe.typeOf[VariableFormat]
+    } else {
+      val cls = value.getClass
+      val runtimeMirror = universe.runtimeMirror(getClass.getClassLoader)
+      val clsSymbol = runtimeMirror.classSymbol(cls)
+     clsSymbol.toType
+    }
+  }
+
+  override def getAttributeValue[T : ClassTag](attribute: AttrValue)
+    (implicit ev: TensorNumeric[T]) : AnyRef = {
+    attribute.getDataType match {
+      case DataType.INT32 => Integer.valueOf(attribute.getInt32Value)
+      case DataType.INT64 => Long.box(attribute.getInt64Value)
+      case DataType.DOUBLE => Double.box(attribute.getDoubleValue)
+      case DataType.FLOAT => Float.box(attribute.getFloatValue)
+      case DataType.STRING => attribute.getStringValue
+      case DataType.BOOL => Boolean.box(attribute.getBoolValue)
+      case DataType.REGULARIZER => RegularizerConverter.getAttributeValue(attribute)
+      case DataType.TENSOR => TensorConverter.getAttributeValue(attribute)
+      case DataType.VARIABLE_FORMAT => VariableFormatConverter.getAttributeValue(attribute)
+      case DataType.INITMETHOD => InitMethodConverter.getAttributeValue(attribute)
+      case DataType.MODULE => ModuleConverter.getAttributeValue(attribute)
+      case DataType.NAME_ATTR_LIST => NameListConverter.getAttributeValue(attribute)
+      case DataType.ARRAY_VALUE => ArrayConverter.getAttributeValue(attribute)
+      case _ => throw new IllegalArgumentException
+        (s"${attribute.getDataType} can not be recognized")
+    }
+  }
+
+  override def setAttributeValue[T : ClassTag](
+    attributeBuilder : AttrValue.Builder, value: Any, valueType : universe.Type = typePlaceHolder)
+    (implicit ev: TensorNumeric[T]): Unit = {
+    // to make it compatible with Java types
+    if (valueType == universe.typeOf[Int] ||
```

Comment:
Prefer to leave as it is as for non-generic type we compare the type , for generic type we compare the string representation. also data types should be quite stable and not changed frequently